internal/buf/pool.go

```diff
@@ -39,59 +39,61 @@ func (b *Buf) Release() {
 		return
 	}
 
-	notify := false
+	atomic.AddInt64(&b.segment.pool.totalReleasedBytes, int64(len(b.Data)))
+	atomic.AddInt64(&b.segment.pool.totalReleasedBuffers, 1)
+
+	b.segment.mu.Lock()
+	defer b.segment.mu.Unlock()
 
 	// best effort compare-and-swap, which will pop the buffer off the stack in its appropriate segment
-	if atomic.CompareAndSwapInt32(&b.segment.nextUnallocated, b.nextUnallocated, b.previousNextUnallocated) {
-		// popped from the stack in the segment
-		notify = true
+	if b.segment.nextUnallocated == b.nextUnallocated {
+		b.segment.nextUnallocated = b.previousNextUnallocated
 	}
 
-	if atomic.AddInt32(&b.segment.allocatedBufCount, -1) == 0 {
+	b.segment.allocatedBufCount--
+	if b.segment.allocatedBufCount == 0 {
 		// last allocated Buf, we can reset 'next' to zero
-		atomic.StoreInt32(&b.segment.nextUnallocated, 0)
-
-		notify = true
-	}
-
-	if notify {
-		// this segment just became free, notify other goroutines doing Allocate()
-		select {
-		case b.segment.pool.segmentReleased <- struct{}{}: // notified
-		default: // nobody is waiting
-		}
+		b.segment.nextUnallocated = 0
 	}
 
 	b.Data = nil
-
 	b.segment = nil
 }
 
-func (s *segment) allocate(count int) (Buf, bool) {
-	n := int32(count)
+func (s *segment) allocate(n int) (Buf, bool) {
+	// quick check using shared lock
+	s.mu.RLock()
+	haveRoom := s.nextUnallocated+n <= len(s.data)
+	s.mu.RUnlock()
 
-	for {
-		// see if we have capacity in this segment
-		v := atomic.LoadInt32(&s.nextUnallocated)
-		if v+n > int32(len(s.data)) {
-			// out of space in this segment
-			return Buf{}, false
-		}
+	if !haveRoom {
+		return Buf{}, false
+	}
 
-		if atomic.CompareAndSwapInt32(&s.nextUnallocated, v, v+n) {
-			atomic.AddInt32(&s.allocatedBufCount, 1)
+	// we likely have space, allocate under exclusive lock
+	s.mu.Lock()
+	defer s.mu.Unlock()
 
-			return Buf{v + n, v, s.data[v : v+n : v+n], s}, true
-		}
+	// see if we have capacity in this segment
+	v := s.nextUnallocated
```

Comment:
nit: maybe rename `v` to `nu` or `next` or even `prev`? Not sure what would make it clearer.

Reply:
done